File: scripts/validate_us_publication.mjs
#!/usr/bin/env node

import { createHash } from 'node:crypto';
import { readFile, readdir, stat } from 'node:fs/promises';
import { join, resolve } from 'node:path';

const ROOT = resolve(process.cwd());
const MANIFEST_PATH = join(ROOT, 'us_state_manifest.json');
const STATES_DIR = join(ROOT, 'states');
const CHANGES_DIR = join(ROOT, 'changes');
const EXPECTED_JURISDICTIONS = 51;
const MIN_NATIONAL_CAMERA_COUNT = 40_000;

function sha256(data) {
  return createHash('sha256').update(data).digest('hex');
}

async function readJSON(path) {
  return JSON.parse(await readFile(path, 'utf8'));
}

function assert(condition, message) {
  if (!condition) throw new Error(message);
}

async function validateEntry(entry) {
  const stateCode = entry.stateCode;
  const datasetPath = join(STATES_DIR, `${stateCode}.json`);
  const changesPath = join(CHANGES_DIR, `${stateCode}.json`);
  const datasetData = await readFile(datasetPath);
  const dataset = JSON.parse(datasetData.toString('utf8'));
  const changes = await readJSON(changesPath);

  assert(dataset.metadata.schemaVersion === 2, `[${stateCode}] unsupported dataset schema`);
  assert(Array.isArray(dataset.cameras), `[${stateCode}] cameras is not an array`);
  assert(dataset.metadata.cameraCount === dataset.cameras.length, `[${stateCode}] metadata count mismatch`);
  assert(entry.cameraCount === dataset.cameras.length, `[${stateCode}] manifest count mismatch`);
  assert(entry.fileSizeBytes === datasetData.length, `[${stateCode}] file size mismatch`);
  assert(entry.sha256 === sha256(datasetData), `[${stateCode}] SHA-256 mismatch`);

  const ids = new Set();
  let flockCount = 0;
  let directionCount = 0;
  let operatorCount = 0;

  for (const camera of dataset.cameras) {
    assert(typeof camera.id === 'string' && camera.id.length > 0, `[${stateCode}] camera missing ID`);
    assert(!ids.has(camera.id), `[${stateCode}] duplicate camera ID ${camera.id}`);
    ids.add(camera.id);

    assert(Number.isFinite(camera.latitude), `[${stateCode}] invalid latitude for ${camera.id}`);
    assert(Number.isFinite(camera.longitude), `[${stateCode}] invalid longitude for ${camera.id}`);
    assert(camera.latitude >= -90 && camera.latitude <= 90, `[${stateCode}] latitude out of range`);
    assert(camera.longitude >= -180 && camera.longitude <= 180, `[${stateCode}] longitude out of range`);
    assert(camera.stateCode === stateCode, `[${stateCode}] camera assigned to ${camera.stateCode}`);

    if (camera.vendor === 'Flock Safety') flockCount += 1;
    if (camera.directionDegrees !== null) directionCount += 1;
    if (camera.operatorName !== null) operatorCount += 1;
  }

  assert(entry.flockCount === flockCount, `[${stateCode}] Flock count mismatch`);
  assert(entry.directionCount === directionCount, `[${stateCode}] direction count mismatch`);
  assert(entry.operatorCount === operatorCount, `[${stateCode}] operator count mismatch`);
  assert(dataset.metadata.flockCount === flockCount, `[${stateCode}] dataset Flock count mismatch`);
  assert(dataset.metadata.directionCount === directionCount, `[${stateCode}] dataset direction count mismatch`);
  assert(dataset.metadata.operatorCount === operatorCount, `[${stateCode}] dataset operator count mismatch`);

  assert(changes.schemaVersion === 1, `[${stateCode}] unsupported changes schema`);
  assert(changes.stateCode === stateCode, `[${stateCode}] changes state mismatch`);
  assert(changes.toVersion === entry.version, `[${stateCode}] changes version mismatch`);
  assert(changes.addedCount === changes.added.length, `[${stateCode}] added count mismatch`);
  assert(changes.removedCount === changes.removed.length, `[${stateCode}] removed count mismatch`);
  assert(changes.changedCount === changes.changed.length, `[${stateCode}] changed count mismatch`);

  return { cameraCount: dataset.cameras.length, flockCount };
}

async function main() {
  const manifest = await readJSON(MANIFEST_PATH);
  assert(manifest.schemaVersion === 1, 'Unsupported manifest schema');
  assert(Array.isArray(manifest.states), 'Manifest states is not an array');

  const stateFiles = (await readdir(STATES_DIR)).filter((name) => name.endsWith('.json'));
  const changeFiles = (await readdir(CHANGES_DIR)).filter((name) => name.endsWith('.json'));
  assert(stateFiles.length === manifest.states.length, 'State file count does not match manifest');
  assert(changeFiles.length >= manifest.states.length, 'Change file count does not match manifest');

  const stateCodes = manifest.states.map((entry) => entry.stateCode);
  assert(new Set(stateCodes).size === stateCodes.length, 'Duplicate state entries in manifest');

  if (manifest.states.length === EXPECTED_JURISDICTIONS) {
    assert(
      manifest.totalCameraCount >= MIN_NATIONAL_CAMERA_COUNT,
      `National camera count below integrity floor: ${manifest.totalCameraCount}`
    );
  }

  let totalCameraCount = 0;
  let totalFlockCount = 0;
  for (const entry of manifest.states) {
    const result = await validateEntry(entry);
    totalCameraCount += result.cameraCount;
    totalFlockCount += result.flockCount;
  }

  assert(manifest.totalCameraCount === totalCameraCount, 'National camera total mismatch');
  assert(manifest.totalFlockCount === totalFlockCount, 'National Flock total mismatch');

  const manifestStats = await stat(MANIFEST_PATH);
  console.log(
    `Validated ${manifest.states.length} jurisdictions, ${totalCameraCount.toLocaleString()} cameras, ` +
      `${totalFlockCount.toLocaleString()} Flock-tagged (${manifestStats.size.toLocaleString()}-byte manifest)`
  );
}

main().catch((error) => {
  console.error(error instanceof Error ? error.stack ?? error.message : String(error));
  process.exitCode = 1;
});

File: pipeline/validate_us_publication.py
#!/usr/bin/env python3
"""Independent integrity validation for the published U.S. camera datasets."""

from __future__ import annotations

import hashlib
import json
import math
from pathlib import Path
from typing import Any

ROOT = Path(__file__).resolve().parents[1]
MANIFEST_PATH = ROOT / "us_state_manifest.json"
STATES_DIR = ROOT / "states"
CHANGES_DIR = ROOT / "changes"
EXPECTED_CODES = {
    "AL", "AK", "AZ", "AR", "CA", "CO", "CT", "DE", "FL", "GA",
    "HI", "ID", "IL", "IN", "IA", "KS", "KY", "LA", "ME", "MD",
    "MA", "MI", "MN", "MS", "MO", "MT", "NE", "NV", "NH", "NJ",
    "NM", "NY", "NC", "ND", "OH", "OK", "OR", "PA", "RI", "SC",
    "SD", "TN", "TX", "UT", "VT", "VA", "WA", "WV", "WI", "WY",
    "DC",
}
MINIMUM_NATIONAL_CAMERA_COUNT = 50_000


def read_json(path: Path) -> Any:
    return json.loads(path.read_text(encoding="utf-8"))


def sha256_hex(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def fail(message: str) -> None:
    raise RuntimeError(message)


def validate_camera(camera: dict[str, Any], state_code: str, seen: set[str]) -> None:
    camera_id = str(camera.get("id") or "")
    if not camera_id:
        fail(f"{state_code}: camera is missing an ID")
    if camera_id in seen:
        fail(f"{state_code}: duplicate camera ID {camera_id}")
    seen.add(camera_id)

    latitude = camera.get("latitude")
    longitude = camera.get("longitude")
    if not isinstance(latitude, (int, float)) or not math.isfinite(latitude) or not -90 <= latitude <= 90:
        fail(f"{state_code}: invalid latitude for {camera_id}")
    if not isinstance(longitude, (int, float)) or not math.isfinite(longitude) or not -180 <= longitude <= 180:
        fail(f"{state_code}: invalid longitude for {camera_id}")

    record_state = camera.get("stateCode")
    if record_state is not None and str(record_state).upper() != state_code:
        fail(f"{state_code}: {camera_id} declares state {record_state}")


def main() -> None:
    if not MANIFEST_PATH.exists():
        fail("us_state_manifest.json is missing")

    manifest = read_json(MANIFEST_PATH)
    if manifest.get("schemaVersion") != 1:
        fail(f"Unsupported manifest schema: {manifest.get('schemaVersion')}")

    entries = manifest.get("states")
    if not isinstance(entries, list):
        fail("Manifest states is not an array")

    by_code = {str(entry.get("stateCode", "")).upper(): entry for entry in entries}
    if set(by_code) != EXPECTED_CODES:
        missing = sorted(EXPECTED_CODES.difference(by_code))
        extra = sorted(set(by_code).difference(EXPECTED_CODES))
        fail(f"Manifest jurisdiction mismatch; missing={missing}, extra={extra}")

    total_cameras = 0
    total_flock = 0
    global_ids: set[str] = set()

    for state_code in sorted(EXPECTED_CODES):
        entry = by_code[state_code]
        dataset_path = STATES_DIR / f"{state_code}.json"
        changes_path = CHANGES_DIR / f"{state_code}.json"
        if not dataset_path.exists():
            fail(f"{state_code}: state dataset is missing")
        if not changes_path.exists():
            fail(f"{state_code}: change feed is missing")

        raw = dataset_path.read_bytes()
        if sha256_hex(raw) != str(entry.get("sha256", "")).lower():
            fail(f"{state_code}: SHA-256 mismatch")
        if len(raw) != entry.get("fileSizeBytes"):
            fail(f"{state_code}: file size mismatch")

        dataset = json.loads(raw)
        metadata = dataset.get("metadata")
        cameras = dataset.get("cameras")
        if not isinstance(metadata, dict) or not isinstance(cameras, list):
            fail(f"{state_code}: invalid dataset structure")
        if metadata.get("schemaVersion") not in (1, 2):
            fail(f"{state_code}: unsupported dataset schema")
        if metadata.get("cameraCount") != len(cameras):
            fail(f"{state_code}: metadata camera count mismatch")
        if entry.get("cameraCount") != len(cameras):
            fail(f"{state_code}: manifest camera count mismatch")

        state_ids: set[str] = set()
        flock_count = 0
        bearing_count = 0
        operator_count = 0
        for camera in cameras:
            validate_camera(camera, state_code, state_ids)
            # Global IDs from enriched datasets can be truly global; compact
            # state index IDs include coordinates and are also expected unique.
            if camera["id"] in global_ids:
                fail(f"Camera ID appears in multiple state files: {camera['id']}")
            global_ids.add(camera["id"])
            if "flock" in str(camera.get("vendor") or "").casefold():
                flock_count += 1
            if camera.get("directionDegrees") is not None:
                bearing_count += 1
            if str(camera.get("operatorName") or "").strip():
                operator_count += 1

        expected_counts = (
            entry.get("flockCount"),
            entry.get("directionCount"),
            entry.get("operatorCount"),
        )
        actual_counts = (flock_count, bearing_count, operator_count)
        if expected_counts != actual_counts:
            fail(f"{state_code}: manifest attribute counts mismatch {expected_counts} != {actual_counts}")
        metadata_counts = (
            metadata.get("flockCount"),
            metadata.get("directionCount"),
            metadata.get("operatorCount"),
        )
        if metadata_counts != actual_counts:
            fail(f"{state_code}: metadata attribute counts mismatch {metadata_counts} != {actual_counts}")

        bounds = entry.get("bounds") or {}
        required_bounds = ("north", "south", "east", "west")
        if any(not isinstance(bounds.get(key), (int, float)) for key in required_bounds):
            fail(f"{state_code}: invalid or missing state bounds")
        if bounds["north"] <= bounds["south"] or bounds["east"] <= bounds["west"]:
            fail(f"{state_code}: inverted state bounds")

        changes = read_json(changes_path)
        if changes.get("stateCode") != state_code:
            fail(f"{state_code}: change feed state mismatch")
        if changes.get("toVersion") != entry.get("version"):
            fail(f"{state_code}: change feed version mismatch")
        if changes.get("addedCount") != len(changes.get("added", [])):
            fail(f"{state_code}: change feed added count mismatch")
        if changes.get("removedCount") != len(changes.get("removed", [])):
            fail(f"{state_code}: change feed removed count mismatch")
        if changes.get("changedCount") != len(changes.get("changed", [])):
            fail(f"{state_code}: change feed changed count mismatch")

        total_cameras += len(cameras)
        total_flock += flock_count

    if total_cameras < MINIMUM_NATIONAL_CAMERA_COUNT:
        fail(f"National publication has only {total_cameras:,} cameras")
    if manifest.get("totalCameraCount") != total_cameras:
        fail("Manifest national camera total does not match state files")
    if manifest.get("totalFlockCount") != total_flock:
        fail("Manifest national Flock total does not match state files")

    print(
        f"Validated {total_cameras:,} cameras across 50 states + DC "
        f"({total_flock:,} Flock-tagged)."
    )


if __name__ == "__main__":
    main()
